### DISCUSSION TITLE
SH1122 OLED I2C display update speed

### DISCUSSION THREAD

**Discussion by jeroen94704** [olikraus/u8g2#2305] [Q&A] [answered]
I got a couple of 256x64 resolution 2" OLED display modules based on the SH1122 controller wired for I2C operation. These displays work fine with u8g2, except their update speed is veeeeery slooooow. I can actually see the display content updating from top to bottom, which takes about half a second. Looking into this I realized these are 4-bit (16-level) grayscale displays, and with that resolution a full screen update is 256x64x4 = 64k bits. Combined with an I2C clock speed of 400 kHz gives a maximum fps of 6. Given the overhead of the I2C bus protocol and the controller protocol I get where that 2 fps I see comes from.

My question is: Is there a way to improve this? Maybe there's an easy way to mod these modules to use SPI instead of I2C? Or maybe they can be made to run at a higher I2C clock speed? Maybe there's a secret 1-bit monochrome mode? If anyone has any tips I'd love to hear them, since they are otherwise really nice displays and they fit my application perfectly, so it would be a shame to have to replace them.

Thanks in advance!

**Comment by olikraus** [comment, marked as answer]
HW reconfig to SPI should be the best option. You can also try setBusClock function to increase bus clock beyond 400kHz. And finally there are some compiler options to increase speed in u8g2.h

Usually I try to use monochrome modes whenever possible, but if I remember correctly, there was no such mode for this controller.

**Reply by jeroen94704** [reply]
Indeed, there doesn't seem to be a monochrome mode (but I figured maybe I just overlooked it), and increasing the clock speed doesn't seem to work. So I guess I'll have to figure out how to convert the module to SPI, and otherwise I'll just have to take my loss and find something else.

Thanks!

**Reply by olikraus** [reply]
Is the display mounted on a PCB? If so you could send the back side of the same and we could look together for the SPI option.

**Reply by jeroen94704** [reply]
That's very generous of you, thanks!

Looking at the SPI version of this module I didn't see an obvious or easy way to do this, since it looks like a completely different circuit. Here's a photo of the module (I2C version) I have, professionally hot-glued in place:
![SH1122_OLED_Module_Backside](https://github.com/olikraus/u8g2/assets/548463/92e6e1a5-4fe8-4408-9104-98304aeb4a02)

**Reply by olikraus** [reply]
good picture. Ok, I understand your problem.

What I found is this product spec: https://item.taobao.com/item.htm?spm=a21n57.1.0.0.4b74523c6Ps3hm&id=565003520764&ns=1&abbucket=0#detail

Looks like your display. So for SPI, we need to connect IM1 (pin 10 of the display flex connector) to GND.
GND itself seems to be at pin 6 of the display flex connector. 

Looking at the PCB it seems to fit, both pin 6 (GND) and pin 7 (VCL) should be connected to GND, which seems to be the case.

The signal and clock lines are the same like I2C. 
A0 line needs is the address select in I2C, so it is probably the right side of the R8 resistor.
I have not checked CS, but it should be low already and could stay as it is.
Then unsolder pin 10 of the flex cable and connect it to GND.

All in all a little bit complicated and some risk is there to damage the display. Probably it is simpler to get the SPI variant.

**Reply by jeroen94704** [reply]
Wait, so in order to use this display in SPI mode all I have to is:

1. Disconnect pin 10 from Vcc and connect it to GND
2. Remove R8 and connect the right-side pad to be used as A0 (Data/Command selection)

?

That's not too bad! I thought it would be much more complicated. I could even connect A0 to GND and use the display in 3-wire SPI mode, if U8G2 supports it (does it?)

Thanks so much for taking the time to look into this!

**Reply by olikraus** [reply]
> I thought it would be much more complicated.

Ideally CS should get connected also, because CS is used to sync the data transmission. Some displays will not work at all if CS is not connected to the Arduino Board. So, to be on the safe side, you would need to identify CS on the flex cable, disconnect CS from GND (according to the spec, it should be connected to GND) and then reconnect CS to your controller.

Using 3-wire SPI should be ok in theory, however
- The u8g2 implementation for 3-wire SPI has been tested only with a limited number of displays, especially it is not tested with the SH1122 display controller (but it would be cool to hear that this works)
- 3-wire SPI will have a lower performance compared to the 4-wire SPI: One extra bit is required and some software overhead is there to map the 9 bit stream into an 8 bit stream for the Arduino SPI object.

> Remove R8 and connect the right-side pad to be used as A0 (Data/Command selection)

Yes, but using the right-side pad is just an assumption, it might be the left-side also (although it unlikely, because the signal comes from the right). 

I personally would suggest to make a full change description for the PCB, if you really would go ahead with this approach.

I also suggest to trace the PCB connections with an ohm-meter to confirm to which pad IM1 is connected.

**Reply by olikraus** [reply]
Added the spec for the flex cable from the above link 

![grafik](https://github.com/olikraus/u8g2/assets/1084992/908c4053-139c-4754-b468-05f8d548120c)

**Reply by jeroen94704** [reply]
Status report:

I have mostly good news. When I connect IM1 to GND and use the left(!!!) pad of R8 to connect to A0, the display actually works! These 2  modifications are quite easy to perform, and once done I get very smooth and fast screen updates. So that's awesome!

However, it seems you were right about needing CS to make this reliable. With CS tied to GND the display works for one microcontroller (Teensy 4.0) after a power-cycle, but never reliably with another (XIAO SAMD21). Unfortunately, the modification for pin 12 of the ribbon is not so easy as for pin 10. I think there is a trace going right underneath the ribbon cable, because pin 12 stays stubbornly connected to GND, even when I cut the trace left of the solder-pad. So I need to figure out how to do this, but the progress so far is really encouraging. 

To be continued.

**Reply by jeroen94704** [reply]
To give an idea of the difference in performance: https://imgur.com/a/k49hkCb

**Reply by olikraus** [reply]
wow... nice. I am impressed that you really started to change the PCB. Good work.

**Reply by jeroen94704** [reply]
I managed to separate pin 12 (CS) of the ribbon cable and connect it to the CS signal. Happy to report that this works, and the display is now fully functional in SPI mode with both microcontrollers. But this is a pretty dicey mod! In the end I (very!) carefully cut the plastic of the ribbon cable on both sides of the pin 12 pad, desoldered it from the track, lifted it up a bit and stuck a piece of tape underneath. Not something I could've done without [my DIY microscope](https://imgur.com/a/mIp1gJ2).

I think I'll record the process of modding the second display, if for no other reason than to dissuade other people from trying this. My take-away is to stay away from I2C displays and stick with SPI models from now on.

But thanks for all your help @olikraus!!

**Reply by olikraus** [reply]
I am impressed you managed to change everything until it fully works. Amazing work. Nice microscope. But yes, same here. Most soldering nowadays requires magnifying glasses
Anyhow, impressive work.

**Reply by jeroen94704** [reply]
Amazingly, I finally got around to making that video: https://youtu.be/Xg0zBOjHA2Y

**Reply by olikraus** [reply]
Thanks for the video, very nice.
